kind/pkg/cluster/config/encoding/encoding.go

```diff
@@ -0,0 +1,127 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package encoding implements apiVersion aware functionality to
+// Marshal / Unmarshal / Load Config
+package encoding
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/test-infra/kind/pkg/cluster/config"
+)
+
+// Load reads the file at path and attempts to load it as a yaml Config
+// after detecting the apiVersion in the file
+// (or defaulting to the current version if none is specified)
+// If path == "" then the default config for the current version is returned
+func Load(path string) (config.Any, error) {
+	if path == "" {
+		return config.New(), nil
+	}
+	// read in file
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	// read in some config version
+	// TODO(bentheelder): we do not use or respect `kind:` at all
+	// possibly we should require something like `kind: "KindConfig"`
+	cfg, err := Unmarshal(contents)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, err
+}
+
+// LoadCurrent is equivalent to Load followed by cfg.ToCurrent()
+func LoadCurrent(path string) (*config.Config, error) {
+	cfg, err := Load(path)
+	if err != nil {
+		return nil, err
+	}
+	return cfg.ToCurrent(), nil
+}
+
+// used to sniff a config for it's api version
+type configOnlyVersion struct {
+	APIVersion string `json:"apiVersion,omitempty"`
+}
+
+// helper to sniff and validate apiVersion
+func detectVersion(raw []byte) (version string, err error) {
+	c := configOnlyVersion{}
+	if err := yaml.Unmarshal(raw, &c); err != nil {
+		return "", err
+	}
+	switch c.APIVersion {
+	// default to the current api version if unspecified, or explicitly specified
+	case config.APIVersion, "":
+		return config.APIVersion, nil
+	}
+	return "", fmt.Errorf("invalid version: %v", c.APIVersion)
+}
+
+// Unmarshal is an apiVersion aware yaml.Unmarshall for config.Any
+func Unmarshal(raw []byte) (config.Any, error) {
+	if raw == nil {
+		return nil, fmt.Errorf("nil input")
+	}
+	// sniff and validate version
+	version, err := detectVersion(raw)
+	if err != nil {
+		return nil, err
+	}
+	// load version
+	var cfg config.Any
+	switch version {
+	case config.APIVersion:
+		cfg = &config.Config{}
+	}
+	err = yaml.Unmarshal(raw, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// Marshal marshals any config with kind and apiVersion header
+func Marshal(cfg config.Any) ([]byte, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("nil input")
+	}
+	var buff bytes.Buffer
+	// write kind, apiVersion header
+	buff.WriteString(`kind: "`)
+	buff.WriteString(config.Kind)
+	buff.WriteString("\"\n")
+	buff.WriteString(`apiVersion: "`)
+	buff.WriteString(cfg.Version())
+	buff.WriteString("\"\n")
```

Comment:
1) yeah we should use a struct instead, will do that
2) this is a bit confusing `config.Kind` is the `const` value named `Kind` in the `config` package rather than a field on the config, `cfg`.